controllers/pipeline/pipeline_adapter.go

```diff
@@ -350,21 +393,29 @@ func (a *Adapter) getAllPipelineRunsForSnapshot(snapshot *applicationapiv1alpha1
 
 // prepareSnapshot prepares the Snapshot for a given application and the updated component (if any).
 // In case the Snapshot can't be created, an error will be returned.
-func (a *Adapter) prepareSnapshot(application *applicationapiv1alpha1.Application, component *applicationapiv1alpha1.Component, newContainerImage string) (*applicationapiv1alpha1.Snapshot, error) {
+func (a *Adapter) prepareSnapshot(application *applicationapiv1alpha1.Application, component *applicationapiv1alpha1.Component, newContainerImage string, newComponentSource *applicationapiv1alpha1.ComponentSource) (*applicationapiv1alpha1.Snapshot, error) {
 	applicationComponents, err := a.getAllApplicationComponents(application)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get all Application Components for Application %s", a.application.Name)
 	}
 
 	var snapshotComponents []applicationapiv1alpha1.SnapshotComponent
 	for _, applicationComponent := range *applicationComponents {
+		applicationComponent := applicationComponent
 		containerImage := applicationComponent.Spec.ContainerImage
+
+		var componentSource *applicationapiv1alpha1.ComponentSource
 		if applicationComponent.Name == component.Name {
 			containerImage = newContainerImage
+			componentSource = newComponentSource
+		} else {
+			// Get ComponentSource for the component which is not built in this pipeline
+			componentSource = a.getComponentSourceFromComponent(&applicationComponent)
```

Comment:
I'm puzzled with logic here, why are we passing `newComponentSource` when we can always get Source from component using  getComponentSourceFromComponent?

This also feel like violation of dependency injection pattern, you are passing dependency via param but at the same time for other components you are doing dynamic fething

Reply:
Resolving, I got explanation of component workflow offline